### PR TITLE
Add support for IPLD prime's budgets feature in selectors

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

Provide absolute limits on links traversed for incoming and outgoing requests

# Implementation

- use budget in Traverser utility (add logic to handle budget for first link load)
- create options for MaxLinksPerOutgoingRequest & MaxLinksPerIncomingRequest
- pass through and create traversak budgets as needed (0 vals = no max links)
- add integration tests & unit tests of traverser behavior

### CODE REVIEW

File: impl/graphsync.go [hunk 1]
```diff
@@ -179,11 +197,11 @@ func New(parent context.Context, network gsnet.GraphSyncNetwork,
...
+	requestManager := requestmanager.New(ctx, asyncLoader, linkSystem, outgoingRequestHooks, incomingResponseHooks, networkErrorListeners, requestQueue, network.ConnectionManager(), gsConfig.maxLinksPerOutgoingRequest)
 	requestExecutor := executor.NewExecutor(requestManager, incomingBlockHooks, asyncLoader.AsyncLoad)
 	responseAssembler := responseassembler.New(ctx, peerManager)
 	peerTaskQueue := peertaskqueue.New()
-	responseManager := responsemanager.New(ctx, linkSystem, responseAssembler, peerTaskQueue, requestQueuedHooks, incomingRequestHooks, outgoingBlockHooks, requestUpdatedHooks, completedResponseListeners, requestorCancelledListeners, blockSentListeners, networkErrorListeners, gsConfig.maxInProgressIncomingRequests, network.ConnectionManager())
+	responseManager := responsemanager.New(ctx, linkSystem, responseAssembler, peerTaskQueue, requestQueuedHooks, incomingRequestHooks, outgoingBlockHooks, requestUpdatedHooks, completedResponseListeners, requestorCancelledListeners, blockSentListeners, networkErrorListeners, gsConfig.maxInProgressIncomingRequests, network.ConnectionManager(), gsConfig.maxLinksPerIncomingRequest)
```
Comment (willscott) [inline impl/graphsync.go:204]: these have gotten pretty unwieldy. we probably want options or some other pattern for these constructors at some point

File: requestmanager/server.go [hunk 2]
```diff
@@ -228,6 +242,10 @@ func (rm *RequestManager) processResponseMessage(p peer.ID, responses []gsmsg.Gr
 	filteredResponses = rm.filterResponsesForPeer(filteredResponses, p)
 	rm.updateLastResponses(filteredResponses)
 	responseMetadata := metadataForResponses(filteredResponses)
+	fmt.Println(responseMetadata)
+	if len(filteredResponses) > 0 {
+		fmt.Println(filteredResponses[0].Status())
+	}
```
Comment (willscott) [inline requestmanager/server.go:248]: remove debug fmt's?

File: responsemanager/querypreparer.go [hunk 3]
```diff
@@ -20,9 +22,10 @@ import (
 )
 
 type queryPreparer struct {
-	requestHooks      RequestHooks
-	responseAssembler ResponseAssembler
-	linkSystem        ipld.LinkSystem
+	requestHooks       RequestHooks
+	responseAssembler  ResponseAssembler
+	linkSystem         ipld.LinkSystem
+	maxLinksPerRequest uint64
```
Comment (willscott) [inline responsemanager/querypreparer.go:28]: can we be clear/consistent about whether 0 means fail immediately or no limit?
In this file it seems to mean 'no limit', but in the test in traverser_test "errors correctly, no budget" link budget is set to 0, and the traversal then fails due to being out of budget.

File: requestmanager/server.go (hunk 2 above)
Comment (hannahhoward) [reply requestmanager/server.go:248]: #oops

File: responsemanager/querypreparer.go (hunk 3 above)
Comment (hannahhoward) [reply responsemanager/querypreparer.go:28]: in means no limit. The traverser test gets a nil value for the budget, meaning no budget, when this value is 0

Comment (willscott) [reply responsemanager/querypreparer.go:28]: i think what i'm asking for is a comment to that effect :)

File: impl/graphsync.go (hunk 1 above)
Comment (hannahhoward) [reply impl/graphsync.go:204]: they're initialized by the code itself, so I'd probably lean towards a struct, but I also want to move this whole package to an internal directory :) but yes for later

File: responsemanager/querypreparer.go (hunk 3 above)
Comment (hannahhoward) [reply responsemanager/querypreparer.go:29]: added

**Review by willscott — APPROVED** [review]

File: go.mod [hunk 4]
```diff
@@ -27,7 +27,7 @@ require (
 	github.com/ipfs/go-peertaskqueue v0.2.0
 	github.com/ipfs/go-unixfs v0.2.4
 	github.com/ipld/go-codec-dagpb v1.3.0
-	github.com/ipld/go-ipld-prime v0.12.0
+	github.com/ipld/go-ipld-prime v0.12.3-0.20210929125341-05d5528bd84e
```
[the remark this answers is not in the document]
Comment (hannahhoward) [reply go.mod:30]: will do on monday

[the remark this answers is not in the document]
Comment (hannahhoward) [reply go.mod:30]: before I commit go-graphsync v0.10.0 final

**Pull request merged**